### PR TITLE
Add value function for torch

### PR DESCRIPTION
**Pull request by yonghyuc**
Implement GaussianMLPValueFunction for PyTorch algorithms.

A value function computes loss and returns the value to an algorithm to update the weights. 
An algorithm has two optimizers, for policy and value function.

### CODE REVIEW

File: src/garage/torch/value_functions/gaussian_mlp_value_function.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
+"""A value function based on a GaussianMLP model."""
+import torch
+from torch import nn
+
+from garage.torch.modules import GaussianMLPModule
+from garage.torch.value_functions.base import ValueFunction
+
+
+class GaussianMLPValueFunction(ValueFunction, GaussianMLPModule):
```
Comment (krzentner) [inline src/garage/torch/value_functions/gaussian_mlp_value_function.py:9]: Please don't inherit from `GaussianMLPModule`. We did this with policies, and it turned out to make it impossible to make the pytorch policies into a class hierarchy without some major hacks or undoing it. Instead, `ValueFunction` should inherit from `torch.nn.module`, and the `GaussianMLPModule` should just be a field.

File: src/garage/torch/value_functions/base.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        name (str): Value function name, also the variable scope.
+
+    """
+
+    def __init__(self, env_spec, name):
+        self._mdp_spec = env_spec
+        self.name = name
+
+    @abc.abstractmethod
+    def get_param_values(self):
```
Comment (krzentner) [inline src/garage/torch/value_functions/base.py:19]: I'm not quite sure we need these methods right now. Let's try not having them, and we can add them later if needed.

File: src/garage/torch/algos/vpg.py [hunk 3]
```diff
@@ -148,41 +157,42 @@ def train_once(self, itr, paths):
         obs_flat = torch.cat(filter_valids(obs, valids))
         actions_flat = torch.cat(filter_valids(actions, valids))
         rewards_flat = torch.cat(filter_valids(rewards, valids))
-        advantages_flat = self._compute_advantage(rewards, valids, baselines)
+        returns_flat = torch.cat(filter_valids(returns, valids))
```
Comment (krzentner) [inline src/garage/torch/algos/vpg.py:160]: You have an extra line here.

File: examples/torch/maml_ppo_half_cheetah_dir.py [hunk 4]
```diff
@@ -11,9 +11,9 @@
 from garage.experiment import LocalRunner, MetaEvaluator
```
Comment (ryanjulian) [inline examples/torch/maml_ppo_half_cheetah_dir.py:11]: don't forget to update all the other torch examples

File: src/garage/torch/algos/maml.py [hunk 5]
```diff
@@ -150,6 +149,32 @@ def train_once(self, runner, all_samples, all_params):
 
         return average_return
 
+    def _train_value_function(self, paths):
```
Comment (ryanjulian) [inline src/garage/torch/algos/maml.py:152]: @naeioi this open the possibility of applying MAML to both the VF and the policy. are you interested in trying this?

File: src/garage/torch/algos/trpo.py [hunk 6]
```diff
@@ -72,7 +86,22 @@ def __init__(self,
...
+                         minibatch_size=minibatch_size,
+                         max_optimization_epochs=max_optimization_epochs)
+
+    def _train(self, obs_flat, actions_flat, rewards_flat, returns_flat,
+               advs_flat):
+        policy_loss = self._train_policy(obs_flat, actions_flat, rewards_flat,
+                                         advs_flat)
+        logger.log('Policy loss: {}'.format(policy_loss))
+
+        batch_dataset = BatchDataset((obs_flat, returns_flat),
```
Comment (ryanjulian) [inline src/garage/torch/algos/trpo.py:99]: why was this necessary for minibatching for the value function, but not the policy?

File: src/garage/torch/algos/vpg.py [hunk 7]
```diff
@@ -22,13 +23,18 @@ class VPG(BatchPolopt):
...
+            for policy. This can be an optimizer type such as
+            `torch.optim.Adam` or a tuple of type and dictionary, where
+            dictionary contains arguments to initialize the optimizer.
+            e.g. `(torch.optim.Adam, {'lr' = 1e-3})`
+        vf_optimizer (Union[type, tuple[type, dict]]): Type of optimizer
+            for value function. This can be an optimizer type such as
+            `torch.optim.Adam` or a tuple of type and dictionary, where
+            dictionary contains arguments to initialize the optimizer.
             e.g. `(torch.optim.Adam, {'lr' = 1e-3})`
-        policy_lr (float): Learning rate for policy parameters.
```
[the remark this answers is not in the document]
Comment (ryanjulian) [reply src/garage/torch/algos/vpg.py:31]: yes. should probably add vf_lr for symmetry. it should default to policy_lr, IMO.

File: src/garage/torch/algos/vpg.py [hunk 8]
```diff
@@ -449,9 +496,10 @@ def process_samples(self, itr, paths):
...
             for path in paths
         ])
-        baselines = torch.stack([
-            pad_to_last(self._get_baselines(path),
-                        total_length=self.max_path_length) for path in paths
+        returns = torch.stack([
+            pad_to_last(path['returns'], total_length=self.max_path_length)
+            for path in paths
         ])
+        baselines = self._value_function.predict(obs)
```
Comment (ryanjulian) [inline src/garage/torch/algos/vpg.py:503]: please remove the predict interface.

the intention of this PR is to remove the fit/predict interface and retrieve the VF predictions by calling `forward` on the VF within the same computation graph as the policy.

File: src/garage/torch/value_functions/base.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+    Args:
+        env_spec (garage.envs.env_spec.EnvSpec): Environment specification.
+        name (str): Value function name, also the variable scope.
+
+    """
+
+    def __init__(self, env_spec, name):
+        self._mdp_spec = env_spec
+        self.name = name
```
[the remark this answers is not in the document]
Comment (ryanjulian) [reply src/garage/torch/value_functions/base.py:16]: but where will we find out what the correct prefix is?

File: src/garage/torch/value_functions/base.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+            returns (torch.Tensor): Acquired returns with shape :math:`(N, )`.
+
+        Returns:
+            torch.Tensor: Calculated negative mean scalar value of
+                objective (float).
+
+        """
+
+    @abc.abstractmethod
+    def predict(self, obs):
```
[the remark this answers is not in the document]
Comment (ryanjulian) [reply src/garage/torch/value_functions/base.py:52]: agreed. the idea in this PR is to just consider the VF as any other NN primitive in garage. in torch, that means you retrieve its values with forward. this will allow us to compute the policy loss with a single feedforward, for instance. it also allows us to apply things like MAML to the VF as well as the policy.

File: src/garage/torch/value_functions/base.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+            obs (torch.Tensor): Observation from the environment
+                with shape :math:`(P, O*)`.
+
+        Returns:
+            torch.Tensor: Calculated baselines given observations with
+                shape :math:`(P, O*)`.
+
+        """
+
+    def log_diagnostics(self, paths):
```
Comment (ryanjulian) [inline src/garage/torch/value_functions/base.py:65]: please remove this. we'll add it if we're sure we need it later.

File: src/garage/torch/value_functions/gaussian_mlp_value_function.py (hunk 1 above)
Comment (ryanjulian) [reply src/garage/torch/value_functions/gaussian_mlp_value_function.py:9]: i agree.

File: src/garage/torch/value_functions/base.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (krzentner) [reply src/garage/torch/value_functions/base.py:16]: I had assumed that whatever is calling `log_diagnostics` would know what the prefix is. Anyways, I don't have a super strong opinion on this point.

**Comment by ryanjulian** [comment]
What do the MuJoCo3M benchmarks look like?

File: src/garage/torch/algos/maml.py [hunk 12]
```diff
@@ -63,7 +61,8 @@ def __init__(self,
         self._meta_evaluator = meta_evaluator
         self._policy = policy
         self._env = env
-        self._value_function = value_function
+        self._value_function = copy.deepcopy(inner_algo._value_function)
+        self._lame_vf_state = self._value_function.state_dict()
```
[the remark this answers is not in the document]
Comment (yonghyuc) [reply src/garage/torch/algos/maml.py:65]: @naeioi

File: src/garage/torch/algos/maml.py [hunk 13]
```diff
@@ -150,6 +149,32 @@ def train_once(self, runner, all_samples, all_params):
...
+        # MAML resets a value function to its initial state before training.
+        self._value_function.load_state_dict(self._lame_vf_state)
+
+        obs = np.concatenate([path['observations'] for path in paths], axis=0)
+        returns = np.concatenate([path['returns'] for path in paths])
+        obs = torch.from_numpy(obs)
+        returns = torch.from_numpy(returns)
+
+        vf_loss = self._value_function.compute_loss(obs, returns)
+        self._inner_algo._vf_optimizer.zero_grad()
```
[the remark this answers is not in the document]
Comment (yonghyuc) [reply src/garage/torch/algos/maml.py:172]: Change to `vf_optimizer`

File: src/garage/torch/algos/vpg.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (yonghyuc) [reply src/garage/torch/algos/vpg.py:31]: fixed

File: src/garage/torch/value_functions/base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (yonghyuc) [reply src/garage/torch/value_functions/base.py:52]: Every value function returns distribution object?
If I remove the `predict` function, I need to call `self._value_function(obs).mean.flatten(02)`
but what if there is a value function that directly returns baseline value instead of distribution object?

File: src/garage/torch/value_functions/base.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
+"""Base class for all baselines."""
+import abc
+
+
+class ValueFunction(abc.ABC):
```
[the remark this answers is not in the document]
Comment (yonghyuc) [reply src/garage/torch/value_functions/base.py:5]: But GaussianMLPValueFunction inherits the `GaussianMLPModule` which inherit the `nn.Module`
Is it okay to inherit same class from two different parents?
Or do I need to make the base class to other directory?

File: src/garage/torch/value_functions/gaussian_mlp_value_function.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+            obs (torch.Tensor): Observation from the environment
+                with shape :math:`(N \dot [T], O*)`.
+            returns (torch.Tensor): Acquired returns with shape :math:`(N, )`.
+
+        Returns:
+            torch.Tensor: Calculated negative mean scalar value of
+                objective (float).
+
+        """
+        dist = self.forward(obs)
```
[the remark this answers is not in the document]
Comment (yonghyuc) [reply src/garage/torch/value_functions/gaussian_mlp_value_function.py:93]: fixed

File: src/garage/torch/value_functions/base.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (yonghyuc) [reply src/garage/torch/value_functions/base.py:16]: Currently we use the `name` field to log the value using `tabular.record`.
You can check the code below.
https://github.com/rlworkgroup/garage/blob/04841e31df14642a975354b306501da7088f3bd9/src/garage/torch/algos/vpg.py#L190

File: src/garage/torch/value_functions/base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (yonghyuc) [reply src/garage/torch/value_functions/base.py:52]: fixed

File: src/garage/torch/value_functions/base.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (yonghyuc) [reply src/garage/torch/value_functions/base.py:5]: fixed

File: src/garage/torch/value_functions/gaussian_mlp_value_function.py (hunk 1 above)
Comment (yonghyuc) [reply src/garage/torch/value_functions/gaussian_mlp_value_function.py:9]: fixed

File: src/garage/torch/value_functions/base.py (hunk 2 above)
Comment (yonghyuc) [reply src/garage/torch/value_functions/base.py:19]: fixed

File: src/garage/torch/algos/trpo.py (hunk 6 above)
Comment (yonghyuc) [reply src/garage/torch/algos/trpo.py:99]: I thought for TRPO, there is the `cg_iters` field for `ConjugateGradientOptimizer` and this works very similar with `max_optimization_epochs`.
Is it okay to use minibatch for ConjugateGradientOptimizer?
If so, it need to use same minibatch dataset but with different optimizing iterations? (`cg_iters` and `max_optimization_epochs`)?

File: src/garage/torch/algos/vpg.py (hunk 8 above)
Comment (yonghyuc) [reply src/garage/torch/algos/vpg.py:503]: fixed

File: src/garage/torch/value_functions/base.py (hunk 11 above)
Comment (yonghyuc) [reply src/garage/torch/value_functions/base.py:65]: fixed

**Comment by yonghyuc** [comment]
This is PPO benchmark result for Mujoco1M (5.5e+5)

HalfCheetah-v2_benchmark
Hopper-v2_benchmark
InvertedDoublePendulum-v2_benchmark
InvertedPendulum-v2_benchmark
Reacher-v2_benchmark
Swimmer-v2_benchmark
Walker2d-v2_benchmark

**Comment by yonghyuc** [comment]
This is TRPO benchmark result for Mujoco1M (5.5e+5)

HalfCheetah-v2_benchmark
Hopper-v2_benchmark
InvertedDoublePendulum-v2_benchmark
InvertedPendulum-v2_benchmark
Reacher-v2_benchmark
Swimmer-v2_benchmark
Walker2d-v2_benchmark

File: src/garage/torch/value_functions/__init__.py [hunk 16]
```diff
@@ -0,0 +1,5 @@
+"""Baselines (value functions) which use NumPy as a numerical backend."""
```
Comment (krzentner) [inline src/garage/torch/value_functions/__init__.py:1]: This docstring should read `"""Value Function which use PyTorch."""` (Or something)

File: src/garage/torch/algos/trpo.py (hunk 6 above)
Comment (ryanjulian) [reply src/garage/torch/algos/trpo.py:102]: you can't use minibatching with ConjugateGradientOptimizer, or rather it would be kind of useless. Minibatching only works with SGD-based optimization algorithms, but a CG optimizer doesn't use SGD to choose parameters.

Comment (yonghyuc) [reply src/garage/torch/algos/trpo.py:102]: Yes, so I used mini-batch only for value_function because default optimizer for value function is adam. When I run benchmark, I got a better result with minibatch and `optimization_epochs`

File: src/garage/torch/value_functions/__init__.py (hunk 16 above)
Comment (yonghyuc) [reply src/garage/torch/value_functions/__init__.py:1]: I missed this one. Thanks

File: src/garage/torch/algos/maml.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (krzentner) [reply src/garage/torch/algos/maml.py:65]: Conceptually it's the same, it's just that "lame" sounds a little unprofessional. Either way is okay, I guess.

File: src/garage/torch/algos/vpg.py (hunk 3 above)
Comment (krzentner) [reply src/garage/torch/algos/vpg.py:164]: You still have an extra line here, as far as I can tell.

Comment (yonghyuc) [reply src/garage/torch/algos/vpg.py:164]: I'm sorry I don't know what you are referring to. Could you clarify what the extra line means?

Comment (yonghyuc) [reply src/garage/torch/algos/vpg.py:164]: I'm sorry I don't know what you are referring to. Could you clarify what the extra line means?

Comment (krzentner) [reply src/garage/torch/algos/vpg.py:164]: Okay, looks like the diff was just glitched on my end. Sorry for the confusion.

File: src/garage/torch/value_functions/base.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (krzentner) [reply src/garage/torch/value_functions/base.py:20]: Okay, it sounds like everyone else finds name fields to be overall useful, so I'm convinced we should keep them.

File: src/garage/torch/algos/maml.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ryanjulian) [reply src/garage/torch/algos/maml.py:65]: it's just confusing. use `old` or `previous`

File: src/garage/torch/algos/ppo.py [hunk 17]
```diff
@@ -42,16 +49,18 @@ class PPO(VPG):
             the mean entropy to the surrogate objective. See
             https://arxiv.org/abs/1805.00909 for more details.
         minibatch_size (int): Batch size for optimization.
-        max_optimization_epochs (int): Maximum number of epochs for update.
+        optimization_epochs (int): Number of epochs for update.
```
Comment (ryanjulian) [inline src/garage/torch/algos/ppo.py:52]: please continue to use max_optimization_epochs. there's no reason this algorithm can't use early stopping in the future.

File: src/garage/torch/algos/trpo.py [hunk 18]
```diff
@@ -36,33 +48,38 @@ class TRPO(VPG):
             dense entropy to the reward for each time step. 'regularized' adds
             the mean entropy to the surrogate objective. See
             https://arxiv.org/abs/1805.00909 for more details.
+        minibatch_size (int): Batch size for optimization.
+        optimization_epochs (int): Number of epochs for update.
```
Comment (ryanjulian) [inline src/garage/torch/algos/trpo.py:52]: max_optimization_epochs

File: src/garage/torch/algos/trpo.py (hunk 6 above)
Comment (ryanjulian) [reply src/garage/torch/algos/trpo.py:102]: what if i use the non-default ConjugateGradientOptimizer? will it still minibatch? wouldn't that be incorrect?

File: src/garage/torch/algos/vpg.py [hunk 19]
```diff
@@ -73,8 +83,8 @@ def __init__(
             use_softplus_entropy=False,
             stop_entropy_gradient=False,
             entropy_method='no_entropy',
-            minibatch_size=None,
-            max_optimization_epochs=1,
+            minibatch_size=100,
+            optimization_epochs=10,
```
Comment (ryanjulian) [inline src/garage/torch/algos/vpg.py:87]: please use max_optimization_epochs

File: src/garage/torch/algos/vpg.py [hunk 20]
```diff
@@ -432,7 +483,7 @@ def process_samples(self, itr, paths):
         for path in paths:
             if 'returns' not in path:
                 path['returns'] = tu.discount_cumsum(path['rewards'],
-                                                     self.discount)
+                                                     self.discount).copy()
```
Comment (ryanjulian) [inline src/garage/torch/algos/vpg.py:486]: why copy this?

File: src/garage/torch/algos/vpg.py [hunk 21]
```diff
@@ -191,8 +205,36 @@ def train_once(self, itr, paths):
...
+        r"""Train the policy and value function with minibatch.
+
+        Args:
+            obs_flat (torch.Tensor): Observation from the environment
+                with shape :math:`(N, O*)`.
+            actions_flat (torch.Tensor): Actions fed to the environment
+                with shape :math:`(N, A*)`.
+            rewards_flat (torch.Tensor): Acquired rewards
+                with shape :math:`(N, )`.
+            returns_flat (torch.Tensor): Acquired returns
```
Comment (ryanjulian) [inline src/garage/torch/algos/vpg.py:219]: why do you need returns from the CPU? we have a function to calculate returns in the GPU.

File: src/garage/torch/algos/vpg.py [hunk 22]
```diff
@@ -191,8 +205,36 @@ def train_once(self, itr, paths):
...
+                with shape :math:`(N, )`.
+            advs_flat (torch.Tensor): Advantage value at each step
```
Comment (ryanjulian) [inline src/garage/torch/algos/vpg.py:221]: why continue to calculate advantages in the CPU? the point of this change is that we an feedforward the value function and calculate the returns and advantages in the GPU.

File: src/garage/torch/algos/vpg.py [hunk 23]
```diff
@@ -191,8 +205,36 @@ def train_once(self, itr, paths):
...
+
+        """
+        dataset = BatchDataset(
+            (obs_flat, actions_flat, rewards_flat, returns_flat, advs_flat),
+            self._minibatch_size)
+
+        for epoch in range(self._optimization_epochs):
+            for obs, actions, rewards, returns, advs in dataset.iterate():
+                policy_loss = self._train_policy(obs, actions, rewards, advs)
+                vf_loss = self._train_value_function(obs, returns)
```
Comment (ryanjulian) [inline src/garage/torch/algos/vpg.py:232]: it's not clear that it's algorithmically valid to modify the value function during policy training. if you are feeding act/obs through the value function to calculate predicted returns, these predicted returns will start using information from the current batch to calculate predicted returns, which is not valid. the baseline should only have information about the batches prior to this one.

where did you get the idea of doing this (vs just optimizing it after the policy)?

File: src/garage/torch/algos/vpg.py (hunk 20 above)
Comment (yonghyuc) [reply src/garage/torch/algos/vpg.py:486]: If not, I got the error
`E       ValueError: some of the strides of a given numpy array are negative. This is currently not supported, but will be added in future releases.`
This is becaue we revert the order of output `[::-1]` in the `discount_cumsum` function.
So I need to copy the array

File: src/garage/torch/algos/vpg.py (hunk 21 above)
Comment (yonghyuc) [reply src/garage/torch/algos/vpg.py:219]: Can you clarify what is the meaning of "from the CPU"?
I just get values from `tensor_utils.discount_cumsum`, flatten the values, and make minibatch.
Do I need to make another function in PyTorch to compute the returns?

File: src/garage/torch/algos/vpg.py (hunk 23 above)
Comment (yonghyuc) [reply src/garage/torch/algos/vpg.py:232]: I don't know this is related to above comments what you post, but I thought every value (advantages, baselines, returns, ...) are computed at the beginning of the training process in the `process_samples` function. So even I update the value_function with the policy, I don't need to re-compute the values(advantages and baselines) and it would be okay.

File: src/garage/torch/algos/vpg.py (hunk 20 above)
Comment (ryanjulian) [reply src/garage/torch/algos/vpg.py:486]: How did it work before? What changed?

File: src/garage/torch/algos/vpg.py [hunk 24]
```diff
@@ -432,7 +483,7 @@ def process_samples(self, itr, paths):
         for path in paths:
             if 'returns' not in path:
```
Comment (ryanjulian) [inline src/garage/torch/algos/vpg.py:484]: in what circumstance is `'returns'` not in the path?

File: src/garage/torch/algos/vpg.py (hunk 20 above)
Comment (ryanjulian) [reply src/garage/torch/algos/vpg.py:486]: Can you use some other numpy function to get a reversed view, other than negative indexing?

When you call `copy()` you break the gradient path. If your returns have a differentiable components, e.g. a differentiable reward augmentation, your augmented rewards will no longer be differentiable.

File: src/garage/torch/algos/vpg.py (hunk 23 above)
Comment (ryanjulian) [reply src/garage/torch/algos/vpg.py:232]: ok second read, this seems probably okay to me as long as `tu.discount_cumsum` is making differentiable operations.

are you sure that if this algorithm were moved to the GPU, that all of these tensors would go with it?

File: src/garage/torch/algos/vpg.py [hunk 25]
```diff
@@ -191,8 +205,36 @@ def train_once(self, itr, paths):
             discount=self.discount)
         return np.mean(undiscounted_returns)
 
-    def _train(self, obs, actions, rewards, advantages):
-        r"""Train the algorithm with minibatch.
+    def _train(self, obs_flat, actions_flat, rewards_flat, returns_flat,
```
Comment (ryanjulian) [inline src/garage/torch/algos/vpg.py:208]: it's strange for this function to take rewards, returns, and advantages, because returns can always be calculated from rewards, and advantages can always be calculated using returns and the value_function

i was confused because we usually reserve `process_samples` for operations which happen in the CPU, and use `train` for operations which occur in the computation graph (e.g. the GPU if you are running in GPU mode). this class blurs the line, and it's difficult to tell where the np.ndarrays stop and the torch.Tensors start.

File: src/garage/torch/algos/vpg.py (hunk 20 above)
Comment (yonghyuc) [reply src/garage/torch/algos/vpg.py:486]: This error is from [here](https://github.com/rlworkgroup/garage/blob/a2207eeff9fe7ff84c9db68202ffa9f0f233ac22/src/garage/torch/algos/vpg.py#L503).
Previously, there is only LinearFeatureBaseline which doesn't have to convert numpy.array to torch.tensor. But now we need to convert numpy.array to torch.tensor for GaussianMLPValueFunction.

Comment (yonghyuc) [reply src/garage/torch/algos/vpg.py:486]: According to this [post](https://discuss.pytorch.org/t/torch-from-numpy-not-support-negative-strides/3663/11), problem is in PyTorch. PyTorch doesn't support numpy.array with negative stride.
For the gradient path, the returns values from the `tu.discount_cumsum` is numpy.array not a torch.tensor, so I think there is no gradient path for the values yet. 

Also I'm just curious that we use rewards from environment to compute the return values and there is module or functions that compute gradient for the rewards?

File: src/garage/torch/algos/vpg.py (hunk 23 above)
Comment (yonghyuc) [reply src/garage/torch/algos/vpg.py:232]: As I understand, I have to convert tensors using CUDA commands like 
```
if torch.cuda.is_availabe():
    tensors = tensors.cuda()
```
Are you talking about this codes?

File: src/garage/torch/algos/vpg.py (hunk 21 above)
Comment (ryanjulian) [reply src/garage/torch/algos/vpg.py:219]: the point here is that the returns can be differentiable.

File: src/garage/torch/algos/maml.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (naeioi) [reply src/garage/torch/algos/maml.py:65]: Sorry I am late to the party. I made this name. This is not an old or previous value function, but a value function without any training. Now I think `initial_vf_state` is the most appropriate name.

File: src/garage/torch/algos/vpg.py (hunk 25 above)
Comment (yonghyuc) [reply src/garage/torch/algos/vpg.py:208]: I'll take out returns and baselines from the `process_samples` and put these in `train_once`.
Because it has to compute `returns` and `advantages` using the whole trajectories, these values should be computed before separating values into mini-batches.

**Comment by yonghyuc** [comment]
I added new OptimizerWrapper class. 
The OptimizerWrapper gets optimizer type(torch.optim.optimizer) and other parameters for mini batch. So, it is similar with torch.optim.optimizer (`zero_grad`, `step`) but it has `get_minibatch` fucntion to provide batch data to algorithm. 

Overall process is
algo gets mini batch dataset from OptimizerWrapper
OptimizerWrapperzero_grad() ->  inner torch.optim.optimizer.zero_grad()
algo computes loss and backward it
OptimizerWrapper.step() -> inner torch.optim.optimizer.step()

File: src/garage/torch/algos/vpg.py [hunk 26]
```diff
@@ -295,8 +332,8 @@ def _compute_advantage(self, rewards, valids, baselines):
 
         if self._center_adv:
             means = advantage_flat.mean()
-            variance = advantage_flat.var()
-            advantage_flat = (advantage_flat - means) / (variance + 1e-8)
+            std = advantage_flat.std()
```
Comment (ryanjulian) [inline src/garage/torch/algos/vpg.py:335]: why this change? it seems unrelated to value functions

**Comment by yonghyuc** [comment]
@naeioi Could you check the changes and how they affect MAML algorithm?
I changed the MAML following the changes but I am worried about missing part. 
You can check main changes [here](https://github.com/rlworkgroup/garage/pull/1317#issuecomment-620270310)
Thanks!

@krzentner Could you check this PR for me? Thanks!

File: src/garage/torch/algos/trpo.py (hunk 6 above)
Comment (ryanjulian) [reply src/garage/torch/algos/trpo.py]: i saw in the CHANGELOG that Torch 1.5 started making it easier to add custom optimizers, but didn't look in detail.

File: src/garage/torch/algos/vpg.py (hunk 23 above)
Comment (ryanjulian) [reply src/garage/torch/algos/vpg.py]: i am asking whether all of these operations are performed as part of the torch computation graph.

File: tests/benchmarks/garage/tf/algos/benchmark_ppo.py [hunk 27]
```diff
@@ -264,49 +247,3 @@ def run_garage_tf(env, seed, log_dir):
         dowel_logger.remove_all()
 
         return tabular_log_file
-
-
-def run_baselines(env, seed, log_dir):
```
Comment (zequnyu) [inline tests/benchmarks/garage/tf/algos/benchmark_ppo.py:269]: Do we want to delete this? This should be the only place we could locate the baselines code.

File: tests/benchmarks/garage/tf/algos/benchmark_trpo.py [hunk 28]
```diff
@@ -218,59 +215,3 @@ def run_garage(env, seed, log_dir):
         dowel_logger.remove_all()
 
         return tabular_log_file
-
-
-def run_baselines(env, seed, log_dir):
```
Comment (zequnyu) [inline tests/benchmarks/garage/tf/algos/benchmark_trpo.py:223]: Same here

File: src/garage/torch/algos/maml.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (naeioi) [reply src/garage/torch/algos/maml.py:65]: Can you change this to `_initial_vf_state `?

File: tests/benchmarks/garage/tf/algos/benchmark_ppo.py (hunk 27 above)
Comment (yonghyuc) [reply tests/benchmarks/garage/tf/algos/benchmark_ppo.py:269]: It looks like the openai.baseline version is not support Tensorflow 2.0.

File: tests/benchmarks/garage/tf/algos/benchmark_trpo.py (hunk 28 above)
Comment (yonghyuc) [reply tests/benchmarks/garage/tf/algos/benchmark_trpo.py:223]: It looks like the openai.baseline version is not support Tensorflow 2.0.

File: src/garage/torch/algos/maml_trpo.py [hunk 29]
```diff
@@ -3,7 +3,8 @@
 
 from garage.torch.algos import _Default, VPG
 from garage.torch.algos.maml import MAML
-from garage.torch.optimizers import ConjugateGradientOptimizer
+from garage.torch.optimizers import ConjugateGradientOptimizer, \
```
Comment (ryanjulian) [inline src/garage/torch/algos/maml_trpo.py:6]: please just make these separate lines. \ are discourages by PEP8

**Comment by ryanjulian** [comment]
@Mergifyio rebase

**Comment by ryanjulian** [comment]
@Mergifyio rebase
